core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -104,11 +104,13 @@ public static UnsafeExternalSorter createWithExistingInMemorySorter(
       int initialSize,
       long pageSizeBytes,
       int numElementsForSpillThreshold,
-      UnsafeInMemorySorter inMemorySorter) throws IOException {
+      UnsafeInMemorySorter inMemorySorter,
+      long existingMemoryConsumption) throws IOException {
     UnsafeExternalSorter sorter = new UnsafeExternalSorter(taskMemoryManager, blockManager,
       serializerManager, taskContext, recordComparatorSupplier, prefixComparator, initialSize,
         pageSizeBytes, numElementsForSpillThreshold, inMemorySorter, false /* ignored */);
     sorter.spill(Long.MAX_VALUE, sorter);
+    taskContext.taskMetrics().incMemoryBytesSpilled(existingMemoryConsumption);
```

Comment:
Why didn't you set this value in the caller side (`UnsafeKVExternalSorter`)?

Reply:
cc @maropu @cloud-fan

Reply:
Is it the same to use `inMemorySorter.getMemoryUsage`?

Also, shall we update `sorter.totalSpillBytes`?

Reply:
Or, if we can do `final long spillSize = freeMemory() + inMemorySorter.getMemoryUsage()` in `sorter.spill()`?

Reply:
> Is it the same to use `inMemorySorter.getMemoryUsage`?
> 
> Also, shall we update `sorter.totalSpillBytes`?

1. It's not the same with `inMemorySorter.getMemoryUsage`.

When we insert a record into `UnsafeExternalSorter`, the record itself is copied to sorter's `allocatedPages`, and stores the memory address into `InMemorySorter`. So when we spill it, the memory size is the sum of `inMemorySorter.getMemoryUsage` 
 and `allocatePages`, that's what it does in `UnsafeExternalSorter.spill`.

But when we do `sorter.spill`  in `UnsafeExternalSorter.createWithExistingInMemorySorter`,  the sorter's `allocatedPages` is empty since we didn't copy any records to it. The memory address in `inMemorySorter` points to the memory pages in `BytesToBytesMap`.

The passed parameter denotes the size of memory pages in `BytesToBytesMap`.

2. yes, seems we also need to update `sorter.totalSpillBytes`

Reply:
I see, thanks for your explanation.